### ISSUE TITLE
PSyclone is producing invalid code for LFRic

### ISSUE THREAD

**Issue by rupertford** [stfc/PSyclone#1315]
@TeranIvy has found that PSyclone master is producing incorrect LFRic code. This obviously needs to be fixed.

**Comment by rupertford** [comment]
I have reproduced the problem in my local environment.

**Comment by rupertford** [comment]
Iva kindly did a diff of the previous working code vs what we get now for one of the algorithm files : `flux_alg_mod.x90`

All differences are in the halo exchanges created in the PSy layer. The first two differences are given below with the use of `max` for `depth` being the incorrect code.

```
125,126c125,126
<       IF (md_w3_proxy%is_dirty(depth=max(flux_stencil_extent,))) THEN
<         CALL md_w3_proxy%halo_exchange(depth=max(flux_stencil_extent,))
---
>       IF (md_w3_proxy%is_dirty(depth=flux_stencil_extent)) THEN
>         CALL md_w3_proxy%halo_exchange(depth=flux_stencil_extent)
129,130c129,130
<       IF (chi_sph_proxy(1)%is_dirty(depth=max(flux_stencil_extent,1))) THEN
<         CALL chi_sph_proxy(1)%halo_exchange(depth=max(flux_stencil_extent,1))
---
>       IF (chi_sph_proxy(1)%is_dirty(depth=flux_stencil_extent)) THEN
>         CALL chi_sph_proxy(1)%halo_exchange(depth=flux_stencil_extent)
```

**Comment by rupertford** [comment]
I confirm that, as reported by @TeranIvy, the error was introduced when #1229 went on master.

#1229 adds a container to the top of an InvokeSchedule. Problems associated with that PR were related to places where it was assumed that InvokeSchedule would be the root node (which it now is not). This could be the cause of the problem.

**Comment by rupertford** [comment]
Created branch [1315_lfric_gen_broken](https://github.com/stfc/PSyclone/tree/1315_lfric_gen_broken)

**Comment by rupertford** [comment]
I think I have found the problem, or at least I have found part of the problem. The new container node that was added as a parent of `invoke_schedule` is the parent of *all* schedules associated with invokes found in the algorithm layer. The `preceding()` and `following()` dependency analysis routines in the `Node` class work out dependencies from `root` which means they are including dependencies between *different* invoke schedules. I think the container node behaviour is correct i.e. we only want one (but still need to convince myself) and assuming so, I need to change the `preceding()` and `following()` routines to only return nodes from within the current schedule. By this I think I mean the current `InvokeSchedule`.
